[clang] Implement gcc_struct attribute on Itanium targets

DanShaders opened the pull request:
This commit implements gcc_struct attribute with the behavior similar to one in GCC. Current behavior is as follows:

When ItaniumRecordLayoutBuilder is used, [[gcc_struct]] will locally cancel the effect of -mms-bitfields on a record. If -mms-bitfields is not supplied and is not a default behavior on a target, [[gcc_struct]] will be a no-op. This should provide enough compatibility with GCC.

If C++ ABI is "Microsoft", [[gcc_struct]] will currently always produce a diagnostic, since support for it is not yet implemented in MicrosoftRecordLayoutBuilder. Note, however, that all the infrastructure is ready for the future implementation.

In particular, check for default value of -mms-bitfields is moved from a driver to ASTContext, since it now non-trivially depends on other supplied flags. This also, unfortunately, makes it impossible to use usual argument parsing for `-m{no-,}ms-bitfields`.

The patch doesn't introduce any backwards-incompatible changes, except for situations when cc1 is called directly with `-mms-bitfields` option.

Work towards #24757

DanShaders commented:
Hmmm, formatting issues are in the code which I did not change.

DanShaders commented:
And, I guess, CC @erichkeane

DanShaders committed:
[clang] Implement gcc_struct attribute on Itanium targets

This commit implements gcc_struct attribute with the behavior similar to
one in GCC. Current behavior is as follows:

When ItaniumRecordLayoutBuilder is used, [[gcc_struct]] will locally
cancel the effect of -mms-bitfields on a record. If -mms-bitfields is
not supplied and is not a default behavior on a target, [[gcc_struct]]
will be a no-op. This should provide enough compatibility with GCC.

If C++ ABI is "Microsoft", [[gcc_struct]] will currently always produce
a diagnostic, since support for it is not yet implemented in
MicrosoftRecordLayoutBuilder. Note, however, that all the infrastructure
is ready for the future implementation.

In particular, check for default value of -mms-bitfields is moved from
a driver to ASTContext, since it now non-trivially depends on other
supplied flags. This also, unfortunately, makes it impossible to use
usual argument parsing for -m{no-,}ms-bitfields.

The patch doesn't introduce any backwards-incompatible changes, except
for situations when cc1 is called directly with -mms-bitfields option.

Work towards #24757

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3635,6 +3635,13 @@ def MSStruct : InheritableAttr {
+def GCCStruct : InheritableAttr {
+  let Spellings = [GCC<"gcc_struct">];
+  let Subjects = SubjectList<[Record]>;
+  let Documentation = [Undocumented];
```
Documentation required.

erichkeane reviewed (commented):
This needs attribute documentation and a release note.

Additionally, I think we should prohibit using this in Microsoft mode (that is, diagnose that) rather than having it be a no-op.  We can enable it in the future, but if we permit it, folks will assume it works.

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -302,6 +302,11 @@ Attribute Changes in Clang
   to reduce the size of the destroy functions for coroutines which are known to
   be destroyed after having reached the final suspend point.
 
+- On targets with C++ ABI other than Microsoft, Clang now supports
```
This should just say 'Itanium' rather than talking around it (other than Microsoft).

erichkeane commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -302,6 +302,11 @@ Attribute Changes in Clang
   be destroyed after having reached the final suspend point.
 
+- On targets with C++ ABI other than Microsoft, Clang now supports
+  ``[[gnu:gcc_struct]]`` with the behavior similar to one existing in GCC. In
```
I don't know what the second half here is trying to say, can you rephrase?

erichkeane commented on clang/include/clang/Basic/DiagnosticASTKinds.td:
```diff
@@ -998,6 +998,9 @@ def warn_npot_ms_struct : Warning<
   DefaultError, InGroup<IncompatibleMSStruct>;
 
+def err_itanium_layout_unimplemented : Error<
+  "sorry, Itanium-compatible layout is unimplemented for the current C++ ABI">;
```
this isn't really a good error format, please try to make this look more like our other diagnostics.

erichkeane reviewed (commented):
Docs need work, code is fine from what I can tell, but @efriedma-quic or @rjmccall need to take a look at the ABI components.

rjmccall commented on clang/lib/AST/Decl.cpp:
```diff
@@ -5031,7 +5031,12 @@ void RecordDecl::completeDefinition() {
+    return true;
+  if (hasAttr<GCCStructAttr>())
+    return false;
+  return C.getLangOpts().MSBitfields.value_or(
```
Is there a reason we don't just set the lang opt correctly for the target?  I'm pretty sure language option defaults depending on target settings is well-established.

DanShaders replied:
But there is no `TargetInfo` object when I set it in `CompilerInvocation::ParseLangArgs`.

rjmccall replied:
You have a target triple, and you have the base CXXABI. What information are you looking to get from the `TargetInfo`?

DanShaders replied:
I still don't quite understand what you want.

Yes, I can compute default value while parsing arguments, but I still will need `TargetInfo::defaultsToMsStruct`, since I use it in SemaDeclCxx.

rjmccall replied:
I always prefer to have less complicated logic in the options types, because I've seen it get out of hand.  I don't feel too strongly about this, though, and I do see how you have uses of the components feeding into it.

rjmccall commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -7255,20 +7255,27 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
+  // Microsoft or Itanium C++ ABI. However, even if these attributes are
+  // present, we do not layout classes following foreign ABI rules, but
+  // instead enter a special "compatibility mode", which only changes
+  // alignements of fundamental types and layout of bit fields.
```
```suggestion
  // alignments of fundamental types and layout of bit fields.
```

DanShaders commented:
@rjmccall Would you mind merging this then? (I don't have write access)

AaronBallman commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -3643,7 +3643,14 @@ def CFGuard : InheritableAttr, TargetSpecificAttr<TargetWindows> {
+def GCCStruct : InheritableAttr {
+  let Spellings = [GCC<"gcc_struct">];
+  let Subjects = SubjectList<[Record]>;
+  let Documentation = [GCCStructDocs];
```
```suggestion
  let Documentation = [MSStructDocs]; // Covers this attribute too.
```

AaronBallman commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -7482,3 +7482,26 @@ generation of the other destruction cases, optimizing the above `foo.destroy` to
+``-mno-ms-bitfields`` switches and via ``#pragma ms_struct``.
+  }];
+}
+
+def GCCStructDocs : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+See ms_struct_.
+  }];
+}
```
```suggestion

```

AaronBallman reviewed (commented):
The attribute bits LGTM, but I added some more reviewers to double-check the driver changes.

rjmccall commented:
I agree with the Sema/AST-level LGTM (but also don't feel comfortable approving the driver changes)

MaskRay commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3618,6 +3618,13 @@ void CompilerInvocationBase::GenerateLangArgs(const LangOptions &Opts,
 
+  if (Opts.MSBitfields.has_value()) {
+    if (Opts.MSBitfields.value())
+      GenerateArg(Consumer, OPT_mms_bitfields);
```
If we use `BoolOption` in `Options.td`, we won't need this custom code.

MaskRay commented on clang/include/clang/Driver/Options.td:
```diff
@@ -4393,15 +4393,15 @@ def : Joined<["-"], "mmacosx-version-min=">,
   Visibility<[ClangOption, CC1Option]>,
-  HelpText<"Set the default structure layout to be compatible with the Microsoft compiler standard">,
-  MarshallingInfoFlag<LangOpts<"MSBitfields">>;
+  HelpText<"Set the default structure layout to be compatible with the Microsoft compiler standard">;
```
The Driver and Frontend changes seem unneeded? `-mno-ms-bitfields` can remain a driver-only option that is not in CC1.

DanShaders replied:
Once again, yes, I can theoretically use `BoolOption` and leave everything as it is. However, this requires computing default value for `-mms-bitfields` in driver, which is possible but requires code duplication and is in general less clean that the current solution. I provided some more details in the reply to rjmccall above.

In case you are strongly against breaking the convention, I can alteranatively make both `-mms-bitfields` and `-mno-ms-bitfields` driver-only options and invent a new enum option for cc1 along the lines of `-flayout-compatiblity-type={default,microsoft,itanium}`. I should note, however, that this would only slightly change the way I pass arguments from the driver to cc1.

DanShaders commented:
@MaskRay, I figured the alternative solution with the completely decoupled driver/frontend flags might actually be better than my initial approach. The only issue I see with it is that it requires `s/-mms-bitfields/-fms-layout-compatibility=microsoft/g` in quite a large number of tests which invoke frontend directly.

Would you mind taking a look at this again?

MaskRay commented:
`-mms-bitfields` is a GCC x86 specific option (`aarch64-linux-gnu-gcc -mms-bitfields -xc /dev/null -E` => `error: unrecognized command-line option ‘-mms-bitfields’`).
It does not make sense for MSVC.
Currently, Driver defaults to pass -mms-bitfields to cc1 for `windows-gnu` triples (https://reviews.llvm.org/D81795).
`windows-msvc` triples don't get `-mms-bitfields`.

This patch changes `getCXXABI().isMicrosoft()` to use the `-mms-bitfields` behavior (`RecordDecl::isMsStruct`), which is a drastic change.
Is it really necessary to achieve your goal?

For the `gcc_struct` feature request #24757 , I believe it's for windows-gnu triples, not for windows-msvc.
We don't necessarily enable it for windows-msvc. The patch tries to do something with windows-msvc and introduces CC1 options whose names are different from Driver options.
These introduce a lot complexity.

MaskRay reviewed (changes requested):
.

DanShaders commented:
@MaskRay 

> It does not make sense for MSVC.

Let me share a bit of background here. While porting SerenityOS's libraries to Windows and, specifically, to `x86_64-pc-windows-msvc`, we discovered a few tests that were mysteriously failing. It turned out that the change in behavior was due to different bit-field layout algorithm used in Itanium and Microsoft C++ ABIs. After consulting the documentation, I figured `-mno-ms-bitfields` should fix this and all future bitfield-related problems on Windows, and the option indeed worked for the reduced testcase I had locally. Unfortunately, it didn't seem to make any difference elsewhere (because `-m{no,}ms-bitfields` is silently ignored for `windows-msvc` but I haven't known that at the time). So, in the end, we were forced to just work around the issue in a very ugly way (see SerenityOS/serenity#21722 and SerenityOS/serenity#21781).

While we probably won't enable `-mno-ms-bitfields` globally even when the support for it will be implemented in `MicrosoftRecordLayoutBuilder`, I believe silently ignoring an option is not an acceptable solution (note that with this PR Clang will emit a diagnostic about unsupported layout compatibility type). Additionally, if one wraps all external (i.e., Windows) headers with `#pragma ms_struct on`, using them while compiling with `-mno-ms-bitfields` will be valid.

> This patch changes getCXXABI().isMicrosoft() to use the -mms-bitfields behavior (RecordDecl::isMsStruct), which is a drastic change.

This won't change anything observable at all. When `getCXXABI().isMicrosoft()` is true, `RecordDecl::isMsStruct` returning true means using default (old) behavior. I specifically checked all its users and ensured this is the case.

> For the gcc_struct feature request https://github.com/llvm/llvm-project/issues/24757 , I believe it's for windows-gnu triples, not for windows-msvc.

As I said, supporting `[[gcc_struct]]` on `windows-msvc` will allow to get rid of hacks we have in Serenity. There is a similar issue in QEMU (https://gitlab.com/qemu-project/qemu/-/issues/1782#note_1495842591), albeit there they were talking about `windows-mingw`. On top of that, in the mentioned issue, Erich Keane seemed to indirectly acknowledge the need for `[[gcc_struct]]` for `windows-msvc`.

> These introduce a lot complexity.

It's not clear to me how to implement this in a simpler way. The requirements were the following:
- I need default value for `-mms-bitfield` available in frontend to use it in SemaCXXDecl.cpp
- I can't add `defaultsToMsStruct` to `llvm::Triple`, since its value also depends on `-fc++abi=` (which is partially ignored for `windows-msvc` but that's another story).
- I don't want to calculate the default value in two places independently.
- I need either to be able to query if `-m{no-,}ms-bitfields` was provided or to compute default value for it in driver.

MaskRay commented:
I am now confused by the subject "[clang] Stub out gcc_struct attribute".
Do you mean "Implement gcc_struct attribute"? But the description isn't clear that this patch changes `-mms-bitfields` and `ms_struct` to be respected for windows-msvc targets.

SerenityOS is an OS, independent from Windows. What does porting it to Windows mean? Build some SerenityOS components on Windows targeting the PE object file format.
I don't know how this works but the test portability seems a weak argument to add an ABI variant for windows-msvc.

I am sure there are many other differences between ELF/PE struct layouts other than those controlled by `-mms-bitfields`/`-mno-ms-bitfields`.
If SerenityOS tests want to be portable with ELF and PE, you likely need to provide two sets of golden values.

---

https://gitlab.com/qemu-project/qemu/-/issues/1782#note_1495842591 seems like ignored `gcc_struct` for windows-gnu targets (feature request https://github.com/llvm/llvm-project/issues/24757).
I agree that if there are real needs and the needs seem genuine, Clang should support `gcc_struct`.

---

`-mms-bitfields` is a GCC x86 specific option. Darwin targets have some work recently in 2018 (https://reviews.llvm.org/D42660), but I am unsure whether 
`-mms-bitfields` wants to be supported by all architectures, or just x86 Darwin targets.
For ELF targets, I think we want to make `-mms-bitfields` x86-specific and issue a `error: unsupported option '-mms-bitfields' for target 'aarch64'`-style error (https://maskray.me/blog/2023-08-25-clang-wunused-command-line-argument)

ADKaster commented:
> SerenityOS is an OS, independent from Windows. What does porting it to Windows mean? Build some SerenityOS components on Windows targeting the PE object file format?

In this case, Dan is referring to future plans to port the Ladybird browser to x86_64-windows-msvc. The browser and its ports use the system libraries from SerenityOS in the implementation.

DanShaders commented:
@MaskRay 

> I am now confused by the subject "[clang] Stub out gcc_struct attribute".

The user-facing change is `[[gcc_struct]]` attribute implemented for Itanium C++ ABI. Better handling of `-mms-bitfields` is just a byproduct. I agree that commit message should be more specific about changes to them.

> `-mms-bitfields` wants to be supported by all architectures, or just x86 Darwin targets.

It's not like supporting `-mms-bitfields` for all targets adds a big maintenance burden. The architecture check you are talking about can be easily added in the future if proves to be needed.

------

As Andrew said, I indeed was referring to code used by cross-platform Ladybird browser.

Microsoft bit-field layout didn't break an overly-specific regression test but rendered unusable double to string conversion. The culprit was the following snippet:
```c++
union Extractor {
  double value;
  struct {
    bool sign : 1;
    u32 exponent : 11;
    u64 mantissa : 52;
  };
};
```
According to MSVC ABI, there should be padding between fields. I hope you agree that this is not an intuitive and expected behavior.

I later found that, in Wasi implementation, we have more bit-fields that rely on the absence of padding which are susceptible to the same problem (they look like `struct { 
bool flag1 : 1; bool flag2 : 1; u32 _unused : 30 };`). To fix them, we have to split single `_unused` field into 3 (and this field cannot be a transformed to a padding since we have to ensure it is zeroed).

mstorsjo commented:
> `-mms-bitfields` is a GCC x86 specific option (`aarch64-linux-gnu-gcc -mms-bitfields -xc /dev/null -E` => `error: unrecognized command-line option ‘-mms-bitfields’`).

While it is implemented as an x86 specific option in GCC right now, that doesn't mean that it only is supposed to have an effect for x86. Upstream GCC only supports Windows on x86, and my recollection is that lots of the Windows specific logic is located in x86 specific files, even if the same logic also should apply for Windows on any other architecture - it's just not implemented (yet).

As implemented in Clang, the option works for any MinGW target.

As an example:
```c
struct field {
        unsigned char a : 4;
        unsigned int b : 4;
};
int size = sizeof(struct field);
```
```console
$ clang -target x86_64-windows-gnu -S -o - bitfields.c | grep -A1 size
        .globl  size                            # @size
        .p2align        2, 0x0
size:
        .long   8                               # 0x8
$ clang -target x86_64-windows-gnu -S -o - bitfields.c -mno-ms-bitfields | grep -A1 size
        .globl  size                            # @size
        .p2align        2, 0x0
size:
        .long   4                               # 0x4
$ clang -target aarch64-windows-gnu -S -o - bitfields.c | grep -A1 size
        .globl  size                            // @size
        .p2align        2, 0x0
size:
        .word   8                               // 0x8
$ clang -target aarch64-windows-gnu -S -o - bitfields.c -mno-ms-bitfields | grep -A1 size
        .globl  size                            // @size
        .p2align        2, 0x0
size:
        .word   4                               // 0x4
```

---

> https://gitlab.com/qemu-project/qemu/-/issues/1782#note_1495842591 seems like ignored `gcc_struct` for windows-gnu targets (feature request #24757).
> I agree that if there are real needs and the needs seem genuine, Clang should support `gcc_struct`.

Yes, this would clearly be good to have. For orthogonality, it would be good to have both `gcc_struct` and `ms_struct` available. GCC does support them on non-windows targets as well; I think that can be useful for implementing things like Wine.

---
As noted somewhere (I don't see the comment to quote right now), the MS bitfield packing logic (as enabled by `-mms-bitfields`) is enabled by default when using the MSVC C++ ABI, but not when using the Itanium C++ ABI. But as referenced, since https://reviews.llvm.org/D81795, we do enable the option `-mms-bitfields` automatically for MinGW targets which otherwise do use the Itanium C++ ABI. Being able to override this back to the default format for individual structs would be great.

I don't know and can't speculate about what the implications would be for being able to switch to GCC struct layout in the MSVC C++ ABI, though. For individual structs, I guess it should be doable (I'm only thinking for trivial-ish structs like the one in my example above, right now though). For anything relating to C++ classes and the mechanisms behind them, I'm pretty sure one doesn't want to change how anything of that works. Therefore I don't think it's too relevant to implement `-mno-ms-bitfields` for MSVC targets (as I guess it could open a huge can of worms).

mstorsjo commented:
> Microsoft bit-field layout didn't break an overly-specific regression test but rendered unusable double to string conversion. The culprit was the following snippet:
> 
> ```c++
> union Extractor {
>   double value;
>   struct {
>     bool sign : 1;
>     u32 exponent : 11;
>     u64 mantissa : 52;
>   };
> };
> ```
> 
> According to MSVC ABI, there should be padding between fields. I hope you agree that this is not an intuitive and expected behavior.

It is indeed an unexpected thing. However it is possible to work around it for all these cases; if you declare the inner structure like this:
```
struct {
  u64 sign : 1;
  u64 exponent : 11;
  u64 mantissa : 52;
};
```
Then there will be no extra padding between the fields.

DanShaders commented:
Yes, I know by now :) But this requires using the same type for all the bit-fields which might lead to unnecessary casts in the algorithm itself. And the other case is not as easy to fix.

DanShaders commented:
> Therefore I don't think it's too relevant to implement -mno-ms-bitfields for MSVC targets (as I guess it could open a huge can of worms).

I want to only alter the layout of explicit fields in the class with the future MSVC `-mno-ms-bitfields` implementation. This is always save to do without worrying much about other parts of the compiler.

I should stress that regardless of original motivation the additional complexity of `-fms-layout-compatibility` makes checks in a few other places easier to follow.

Also, do we want to support `--target=x86_64-pc-windows-msvc -fc++-abi=itanium -mms-bitfields`? If so, it would be much simpler with my approach.

DanShaders commented:
One more thing. Re binary compatibility concerns: `-mno-ms-bitfields` on MinGW is an equally-sized footgun as on MSVC. Without proper header annotation with `#pragma ms_struct on`, either of them will silently make an ABI mismatch. However, for some reason, supporting `-mno-ms-bitfields` on MinGW is not argued upon.

mstorsjo commented:
> One more thing. Re binary compatibility concerns: `-mno-ms-bitfields` on MinGW is an equally-sized footgun as on MSVC. Without proper header annotation with `#pragma ms_struct on`, either of them will silently make an ABI mismatch. However, for some reason, supporting `-mno-ms-bitfields` on MinGW is not argued upon.

I guess this is for historical reasons. Originally the MinGW target had this as an opt-in, and this was indeed a silent ABI mismatch. Users who needed to use affected structs and interact with Microsoft APIs had to remember to build their code with `-mms-bitfields` (and hope they don't link in some code that is built without it, with affected structs in their interface). It became the default only by GCC 4.7 (in 2012), and we picked up on this way much later in Clang, in Clang 11 in 2020.

DanShaders commented:
Okay, @mstorsjo @MaskRay, what is the way forward?

Am I right that, as for the user-facing changes, `[[gcc_struct]]` cancelling implicit `-mms-bitfilds` on MinGW is fine and silently ignoring `-m{no-}ms-bitfields` on `windows-msvc` is not? When exactly should we disallow `-m{no-,}ms-bitfields`? Should the aforementioned `--target=x86_64-pc-windows-msvc -fc++-abi=itanium -mms-bitfields` be accepted? Is it fine to provide `[[gcc_struct]]` on MSVC because of reasons I outlined before?

mstorsjo commented:
> Okay, @mstorsjo @MaskRay, what is the way forward?

I'm totally not authoritative for these things, but...

> Am I right that, as for the user-facing changes, `[[gcc_struct]]` cancelling implicit `-mms-bitfilds` on MinGW is fine

Sounds quite fine for me

> and silently ignoring `-m{no-}ms-bitfields` on `windows-msvc` is not?

Silently ignoring options is clearly not good IMO, so either we warn about them or implement them

> Should we (and if yes, when exactly) disallow `-m{no-,}ms-bitfields`? Should the aforementioned `--target=x86_64-pc-windows-msvc -fc++-abi=itanium -mms-bitfields` be accepted?

FWIW I wasn't even aware that it was possible to pick a nondefault C++ ABI, so I don't have a strong opinion on this matter. If it works and doesn't create inconsistencies, then I don't mind, but I guess the regular Clang maintainers have more of a final say on that.

> Is it fine to provide `[[gcc_struct]]` on MSVC because of the reasons I outlined before?

I would be totally fine with that.

rjmccall commented:
> @rjmccall The problem will arise only if GCC implements support for MSVC C++ ABI and decides that there is a better way to implement `gcc_struct`. Since, AFAIC, MSVC-compatibility for GCC is not even planned, it's unlikely anybody there will have strong opinions on this. Yet I posted a question on gcc mailing list (https://gcc.gnu.org/pipermail/gcc/2023-December/242963.html) and, unsurprisingly, got no replies in a week.

My question is actually more about supporting `ms_struct` and `gcc_struct` on other architectures than x86 than it's about MSVC-compatible targets.  I agree that it's highly unlikely that GCC would ever implement the MSVC C++ ABI, but e.g. MinGW on ARM doesn't seem out of the picture.

rjmccall commented:
Right, I'd just like to make sure that we're not deepening a divergence here.  It would be good to get agreement from the GCC devs that they think `ms_struct` probably ought to do something on e.g. ARM MinGW targets and that they consider this a bug (in a feature that they may not really support, which is fine).  But if they think *we're* wrong and that this really should only have effect on x86, I would like to know that.

DanShaders commented:
@MaskRay Bump

DanShaders commented:
@rjmccall Re asking GCC developers about gcc_struct on non-x86: https://gcc.gnu.org/pipermail/gcc/2024-January/243154.html. Either GCC devs aren't really worried about this or I can't properly write emails (what's totally possible). 

@MaskRay Any chance you can look at this PR again? Will you be more happy if instead of `-fms-layout-compatibility` there will be a copy of  `TargetInfo::defaultsToMsStruct` in the driver?

rjmccall commented:
> @rjmccall Re asking GCC developers about gcc_struct on non-x86: https://gcc.gnu.org/pipermail/gcc/2024-January/243154.html. Either GCC devs aren't really worried about this or I can't properly write emails (what's totally possible).

Alright, well, we tried.  I think rolling forward with what your proposal on non-x86 platforms is reasonable.

mstorsjo commented:
> > @rjmccall Re asking GCC developers about gcc_struct on non-x86: https://gcc.gnu.org/pipermail/gcc/2024-January/243154.html. Either GCC devs aren't really worried about this or I can't properly write emails (what's totally possible).
> 
> Alright, well, we tried. I think rolling forward with what your proposal on non-x86 platforms is reasonable.

Yes, I don't foresee any issue with it. There's a proposed patchset to GCC for adding support for AArch64 on Windows, but it's all very early. So that shouldn't be taken as reference for how finer details should work in the end anyway. Even if they don't support these attributes from the start, the probably should/will in the end.

DanShaders commented:
Ping @MaskRay


Changes in the force-push:
- rebased on top of main;
- moved `defaultsToMsStruct` to `ASTContext` since it also depends on auxiliary target.

mstorsjo commented:
> Ping @MaskRay
> 
> Changes in the force-push:
> 
> * rebased on top of main;
> * moved `defaultsToMsStruct` to `ASTContext` since it also depends on auxiliary target.

Please update the subject of the PR - this doesn't stub out the attribute, it implements it, if I understand things correctly?

DanShaders commented:
Fair enough

mstorsjo reviewed (commented):
I had a brief look over this, and it looks reasonable to me, although I'm not anywhere near competent enough to properly review it from the Clang perspective and approve it.

I don't see any added test that actually exercises the new `gcc_struct` attribute anywhere though - I'd like to see that, in a test that runs with `-mms-bitfields` or similar, and shows what effect the attribute has in that context.

pbo-linaro commented:
Hi all, any update on this PR?
QEMU still actively block compiling with clang on windows for this reason, by failing at configure step, and it was decided to wait for official `gcc_struct` support in llvm.

DanShaders committed:
fixup! Add test for gcc_struct

mstorsjo commented:
Why the change from `-mms-bitfields` to `-fms-layout-compatibility=microsoft` in tests?

DanShaders commented:
Short story: because otherwise -mno-ms-bitfields gets silently ignored for x86_64-pc-windows-msvc and I want clang to be very explicit about ignoring -mno-ms-bitfields. The user-facing flags remain the same (`-m{no-,}ms-bitfields`), the only time one ever encounters `-fms-layout-compatibility=` is when directly invoking frontend which we happen to do in tests.

More details [here](https://github.com/llvm/llvm-project/pull/71148#discussion_r1407329444) and in the following comment.

mstorsjo commented:
Ah, I see, I missed that the changed tests were on the `-cc1` level.

efriedma-quic commented on clang/test/CodeGen/gcc_struct.c:
```diff
@@ -0,0 +1,18 @@
+    int a : 24;
+    char b : 8;
+} __attribute__((gcc_struct)) t1;
+static int a1[(sizeof(t1) == 4) - 1];
```
static_assert

DanShaders replied:
That's C and ms_struct tests are written this way as well.
Do you want me to convert the file to C++, use _Static_assert, or set standard to C23?

efriedma-quic replied:
Any of those choices is fine.  If I had to pick, I'd say just use _Static_assert.

efriedma-quic commented on clang/include/clang/Basic/DiagnosticASTKinds.td:
```diff
@@ -997,6 +997,9 @@ def warn_npot_ms_struct : Warning<
   DefaultError, InGroup<IncompatibleMSStruct>;
 
+def err_itanium_layout_unimplemented : Error<
+  "Itanium-compatible layout for the Microsoft C++ ABI is not yet supported">;
```
Missing testcase for this diagnostic.

DanShaders committed:
tests for todo diagnostic & static_assert

DanShaders commented:
@MaskRay 7th ping :)

efriedma-quic reviewed (approved):
LGTM

MaskRay commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -6113,9 +6113,13 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
-  if (Args.hasFlag(options::OPT_mms_bitfields, options::OPT_mno_ms_bitfields,
-                   Triple.isWindowsGNUEnvironment())) {
-    CmdArgs.push_back("-mms-bitfields");
+  if (Args.hasArg(options::OPT_mms_bitfields) ||
```
If the LangOption default varies across targets, and you want to forward the driver option to clang -cc1,

```
if (auto *A = Args.getLastArg(..., ...)) {
  if (A->matches(...)) CmdArgs.push_back...
  else CmdArgs.push_back...
}
```

lygstate replied:
@DanShaders Sorry for the bother, if really do not have time to continue it, I'd like to continue it with new MR>

lygstate commented:
mesa also need this
```
./../src/compiler/glsl_types.c:749:8: warning: unknown attribute 'gcc_struct' ignored [-Wunknown-attributes]
  749 | struct PACKED explicit_matrix_key {
      |        ^~~~~~
../../src/util/macros.h:214:35: note: expanded from macro 'PACKED'
  214 | #    define PACKED __attribute__((gcc_struct,__packed__))
```

lygstate commented:
@DanShaders do you still have time on this?

mstorsjo committed:
Try to address https://github.com/llvm/llvm-project/pull/71148#discussion_r1929617748

erichkeane reviewed (commented):
I'd like mildly better documentation.  Otherwise, what is the behavior if someone specifies BOTH `ms_struct` AND `gcc_struct`?  It seems to me we should both test and define what that means.

At the moment, it seems like this would just default to the `gcc_struct`, which seems wrong to me.

MaskRay reviewed (commented):
I want to withdraw  my "Request changes", but github doesn't make it easy.

(
File changed -> Review changes -> Comment (instead of "Request changes" or "Approve").. Hmm, it still shows as "request changes"

I clicked "Re-request review", then the "Requested change" icon disappeared. However, after I remove my entry in the reviewer list, the "Requested change" icon appears again...
)

mstorsjo commented:
Thanks for looking at this!

> Otherwise, what is the behavior if someone specifies BOTH `ms_struct` AND `gcc_struct`? It seems to me we should both test and define what that means.

On one hand, I'm not entirely sure if we want to explicitly document the behaviour for this case - should we rather say that specifying both is unspecified? Then we can make that into an error (due to the user passing directly conflicting attributes) later if we want to?

> At the moment, it seems like this would just default to the `gcc_struct`, which seems wrong to me.

What would be the right behaviour to you? If both are passed, I don't really see which outcome would be less wrong, other than erroring out?

I'll try to expand the documentation briefly about what the differences are.

mstorsjo committed:
doc: Extend the docs about ms_struct/gcc_struct about the main differentiator

mstorsjo commented:
Oh, and also - for someone not very familiar with the Clang codebase at this level; where would the right place be to emit such a diagnostic?

By looking at the diff of the PR, the main place I'd see where interacting with these attributes, is here:
```c++
bool RecordDecl::isMsStruct(const ASTContext &C) const {
  if (hasAttr<GCCStructAttr>())
    return false;
  if (hasAttr<MSStructAttr>())
    return true;
  auto LayoutCompatibility = C.getLangOpts().getLayoutCompatibility();
  if (LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Default)
    return C.defaultsToMsStruct();
  return LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Microsoft;
}
```
Should we add an `if (hasAttr<GCCStructAttr>() && hasAttr<MSStructAttr>()) diagnostic();` here, or do we need to deal with that on an entirely different level somewhere?

erichkeane commented:
> Oh, and also - for someone not very familiar with the Clang codebase at this level; where would the right place be to emit such a diagnostic?
> 
> By looking at the diff of the PR, the main place I'd see where interacting with these attributes, is here:
> 
> ```c++
> bool RecordDecl::isMsStruct(const ASTContext &C) const {
>   if (hasAttr<GCCStructAttr>())
>     return false;
>   if (hasAttr<MSStructAttr>())
>     return true;
>   auto LayoutCompatibility = C.getLangOpts().getLayoutCompatibility();
>   if (LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Default)
>     return C.defaultsToMsStruct();
>   return LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Microsoft;
> }
> ```
> 
> Should we add an `if (hasAttr<GCCStructAttr>() && hasAttr<MSStructAttr>()) diagnostic();` here, or do we need to deal with that on an entirely different level somewhere?

See: https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/Attr.td#L769C7-L769C23.

AaronBallman commented:
> Oh, and also - for someone not very familiar with the Clang codebase at this level; where would the right place be to emit such a diagnostic?
> 
> By looking at the diff of the PR, the main place I'd see where interacting with these attributes, is here:
> 
> ```c++
> bool RecordDecl::isMsStruct(const ASTContext &C) const {
>   if (hasAttr<GCCStructAttr>())
>     return false;
>   if (hasAttr<MSStructAttr>())
>     return true;
>   auto LayoutCompatibility = C.getLangOpts().getLayoutCompatibility();
>   if (LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Default)
>     return C.defaultsToMsStruct();
>   return LayoutCompatibility == LangOptions::LayoutCompatibilityKind::Microsoft;
> }
> ```
> 
> Should we add an `if (hasAttr<GCCStructAttr>() && hasAttr<MSStructAttr>()) diagnostic();` here, or do we need to deal with that on an entirely different level somewhere?

I'd recommend adding something like this to Attr.td:
```
def : MutualExclusions<[MSStruct, GCCStruct]>;
```
and the rest should happen automagically.

mstorsjo committed:
Make gcc_struct and ms_struct conflict with each other

AaronBallman commented:
> Thanks! That seems to work great!
> 
> However - that uncovers a case in the existing tests in this PR, which starts failing:
> 
> ```
> #pragma ms_struct on
> struct {  
>     int a : 24;
>     char b : 8;
> } __attribute__((gcc_struct)) t2;
> _Static_assert(sizeof(t2) == 4, "");
> #pragma ms_struct off      
> ```
> 
> This now also hits the same error with incompatible attributes. This is a case I think should be supported; just like `-mms-bitfields` or `-mno-ms-bitfields`, such a pragma sets the default over a region, while one may want to override the default for specific structs.

Oof. I didn't realize we had that as a pragma. :-/ Yeah, I think the expected priority is: local takes precedence over pragma takes precedence over command line. So you may have to do the exclusion the hard way (manually).

mstorsjo commented:
> Oof. I didn't realize we had that as a pragma. :-/ Yeah, I think the expected priority is: local takes precedence over pragma takes precedence over command line. So you may have to do the exclusion the hard way (manually).

Thanks! Yes, such an order of precedence seems reasonable.

As the pragma seems to show up in the form of an attribute at the level of `AST/Decl.cpp`, at what level should I check this to be able to distinguish pragma+attribute vs attribute+attribute?

erichkeane commented:
> > Oof. I didn't realize we had that as a pragma. :-/ Yeah, I think the expected priority is: local takes precedence over pragma takes precedence over command line. So you may have to do the exclusion the hard way (manually).
> 
> Thanks! Yes, such an order of precedence seems reasonable.
> 
> As the pragma seems to show up in the form of an attribute at the level of `AST/Decl.cpp`, at what level should I check this to be able to distinguish pragma+attribute vs attribute+attribute?

I haven't spent any time with that pragma, but best I can tell is that the pragma creates it (SemaAttr.cpp ~92) as 'implicit'.  

I DO wonder if we should have some sort of fix to the `MutualExclusion` logic as table-gen'ed to allow `explicit > implicit`.  There is some additional work then to make sure that the explicit wins in codegen in this case (OR, we have it remove the implict/skip the implicit add at that point?) so that only 1 shows up in the AST.

Otherwise, you'd need a custom `handleXXXAttr` (remove the `hasSimpleHandler` from Attr.td, then add handlers to SemaDeclAttr.cpp) that do the checks.

Obviously the 'implicit' one is added in a different step, so that probably should check to make sure it isn't adding if the GCC one already is though.

mstorsjo committed:
Manually reimplement MutualExclusion

mstorsjo committed:
Don't emit diagnostics if MSStructAttr is implicit

mstorsjo committed:
Merge branch 'main' into gcc-struct-rebased

mstorsjo commented:
Thanks for the assistance here! I'll go ahead and merge this one soon then, after the CI completes.

pbo-linaro commented:
Congrats for finally merging this! :)

mstorsjo commented:
It turns out that this change did break the `Sema/struct-packed-align.c` testcase when built in mingw mode (see https://github.com/mstorsjo/llvm-mingw/actions/runs/20208875331/job/58011647890); this can be observed in any build if the testcase is amended like this:
```diff
diff --git a/clang/test/Sema/struct-packed-align.c b/clang/test/Sema/struct-packed-align.c
index d6d0724da49f..c996a58c11e4 100644
--- a/clang/test/Sema/struct-packed-align.c
+++ b/clang/test/Sema/struct-packed-align.c
@@ -1,5 +1,6 @@
 // RUN: %clang_cc1 %s -fsyntax-only -verify
 // RUN: %clang_cc1 %s -fsyntax-only -triple=x86_64-windows-coff -verify
+// RUN: %clang_cc1 %s -fsyntax-only -triple=x86_64-windows-gnu -verify
 // RUN: %clang_cc1 %s -fsyntax-only -triple=x86_64-scei-ps4 -verify
 // RUN: %clang_cc1 %s -fsyntax-only -triple=x86_64-sie-ps5 -verify
 
```

With this change, we're getting the following new failures:
```
# | error: 'expected-error' diagnostics seen but not expected:
# |   File /home/martin/code/llvm-project/clang/test/Sema/struct-packed-align.c Line 140: 'n1' declared as an array with a negative size
# |   File /home/martin/code/llvm-project/clang/test/Sema/struct-packed-align.c Line 141: 'n2' declared as an array with a negative size
# |   File /home/martin/code/llvm-project/clang/test/Sema/struct-packed-align.c Line 175: 'o1' declared as an array with a negative size
# | 3 errors generated.
```

Among other, this part of the testcase now behaves differently:
```c
typedef long long  __attribute__((aligned(1))) nt;

struct nS {
  char buf_nr;
  nt start_lba;
};

#if defined(_WIN32) && !defined(__declspec) // _MSC_VER is unavailable in cc1.
// Alignment doesn't affect packing in MS mode.
extern int n1[sizeof(struct nS) == 16 ? 1 : -1];
extern int n2[__alignof(struct nS) == 8 ? 1 : -1];
#else
extern int n1[sizeof(struct nS) == 9 ? 1 : -1];
extern int n2[__alignof(struct nS) == 1 ? 1 : -1];
#endif
```

DanShaders commented:
Hmm, it looks like the previous behavior was a bug: with `x86_64-windows-gnu`, clang should be in `-mms-bitfields` mode (with `sizeof(struct nS) == 16`) but we don't choose first branch either with or without the patch.
